- 关键词 Promise executor exports self value reason status resolve reject then onFulfilled onRejected onResolvedCallbacks onRejectedCallbacks 
- 注意事项
  - new Promise中可以夹杂着异步逻辑
  - 同一个实例可以多次then
```
// 文件： promise-apply.js
let Promise = require('./promise');
//1) new Promise中可以夹杂着异步罗技
//2) 同一个实例可以多次then
let p = new Promise(function(resolve,reject){
    setTimeout(function(){
        resolve('情人节到了');
    },1000)
});
p.then((value)=>{
    console.log('success',value);
},(reason)=>{
    console.log('error',reason);
})
p.then((value)=>{
    console.log('success',value);
},(reason)=>{
    console.log('error',reason);
})
console.log('end');
// 文件：promise.js
function Promise (executor){
    // 在promise内部定义一个状态 当前promise的状态
    let self = this;
    self.value = undefined;
    self.reason = undefined
    self.status = 'pending'; // 默认promise的状态是pengding
    self.onResolvedCallbacks = []; // 存放所有成功的回调
    self.onRejectedCallbacks = []; // 存放所有失败的回调
    function resolve(value){
        if(self.status === 'pending'){
            self.value = value;
            self.status = 'resolved'; // 成功态
            self.onResolvedCallbacks.forEach(fn=>fn());
        }
    }
    function reject(reason){
        if(self.status === 'pending'){
            self.reason = reason;
            self.status = 'rejected'; // 失败态
            // 发布
            self.onRejectedCallbacks.forEach(fn =>fn());
        }
    }
    executor(resolve,reject);
}
Promise.prototype.then = function(onFulfilled,onRejected){
    let self = this;
    if(self.status === 'resolved'){
        onFulfilled(self.value);
    }
    if(self.status === 'rejected'){
        onRejected(self.reason);
    }
    if(self.status == 'pending'){
        // 订阅
        self.onResolvedCallbacks.push(function(){
            onFulfilled(self.value);
        });
        self.onRejectedCallbacks.push(function(){
            onRejected(self.reason);
        });
    }
}
module.exports = Promise;

```